### PR TITLE
Correct the paths of submodules from the include! macro

### LINKED ISSUE #7846: Include! macro behaviors differently between rust-analyzer and standard library
Hello! Recently, I found that seems there are some undocumented magic in the `include!` macro.

With this directory structure:
```
├── Cargo.toml
├── magic
│   ├── lib.rs
│   └── magic.rs
├── src
│   └── main.rs
```

... and those files:
```rust
// magic/lib.rs
pub mod magic;
```

```rust
// magic/magic.rs
pub fn hello() {
    println!("hello, world!")
}
```

```rust
// main.rs
include!("../magic/lib.rs");

fn main() {
    magic::hello();
}
```

Maybe a little strange, these compiles:

image

... and `rustc` told us, those macro was expanded to:

```
#![feature(prelude_import)]
#[prelude_import]
use std::prelude::v1::*;
#[macro_use]
extern crate std;

pub mod magic {
    pub fn hello() {
        {
            ::std::io::_print(::core::fmt::Arguments::new_v1(&["hello, world!\n"],
                                                             &match () {
                                                                  () => [],
                                                              }));
        }
    }
}
fn main() { magic::hello(); }
```

... but `rust-analyzer` gave another version of the `include!` macro expansion, with an error `unresolved module`:

image

RA version:
image

### PR DESCRIPTION
**Pull request by sticnarf**
This PR should fix #7846. It mostly follows the instructions from @edwin0cheng in that issue.

### CODE REVIEW

File: crates/hir_def/src/nameres/mod_resolution.rs [hunk 1]
```diff
@@ -70,8 +71,13 @@ impl ModDir {
                 candidate_files.push(self.dir_path.join_attr(attr_path, self.root_non_dir_owner))
             }
             None => {
-                candidate_files.push(format!("{}{}.rs", self.dir_path.0, name));
-                candidate_files.push(format!("{}{}/mod.rs", self.dir_path.0, name));
+                if is_include_macro {
```
Comment (edwin0cheng) [inline crates/hir_def/src/nameres/mod_resolution.rs:74]: Is is possible to move `file_id.is_include_macro(db.upcast());` inside here , such that :
```rust
if file_id.is_include_macro(db.upcast()) {
```

**Review by edwin0cheng — COMMENTED** [review]
LGTM except a little nit

File: crates/hir_def/src/nameres/mod_resolution.rs (hunk 1 above)
Comment (sticnarf) [reply crates/hir_def/src/nameres/mod_resolution.rs]: Ok, I moved it inside.

**Comment by edwin0cheng** [comment]
Thanks a lot !

bors r+

**Comment by lnicola** [comment]
changelog fix (first contribution) fix module resolution in `include!`ed files

**Pull request merged**